crates/starknet/src/blocks/mod.rs

```diff
@@ -164,13 +210,329 @@ impl HashProducer for StarknetBlock {
 #[cfg(test)]
 mod tests {
     use starknet_api::block::{BlockHash, BlockHeader, BlockNumber, BlockStatus};
-    use starknet_rs_core::types::BlockId;
+    use starknet_rs_core::types::{BlockId, BlockTag};
+    use starknet_types::felt::Felt;
     use starknet_types::traits::HashProducer;
 
     use super::{StarknetBlock, StarknetBlocks};
     use crate::state::state_diff::StateDiff;
     use crate::traits::HashIdentified;
 
+    #[test]
+    fn block_number_from_block_id_should_return_correct_result() {
+        let mut blocks = StarknetBlocks::default();
+        let mut block_to_insert = StarknetBlock::create_pending_block();
+
+        // latest/pending block returns none, because collection is empty
+        assert!(
+            blocks
+                .block_number_from_block_id(BlockId::Tag(starknet_rs_core::types::BlockTag::Latest))
+                .is_none()
+        );
+        assert!(
+            blocks
+                .block_number_from_block_id(BlockId::Tag(
+                    starknet_rs_core::types::BlockTag::Pending
+                ))
+                .is_none()
+        );
+
+        let block_hash = block_to_insert.generate_hash().unwrap();
+        block_to_insert.header.block_number = BlockNumber(10);
+        block_to_insert.header.block_hash = block_hash.into();
+
+        blocks.insert(block_to_insert, StateDiff::default());
+
+        // returns block number, even if the block number is not present in the collection
+        assert!(blocks.block_number_from_block_id(BlockId::Number(11)).is_none());
+        assert!(blocks.block_number_from_block_id(BlockId::Number(10)).is_some());
+        // returns none because there is no block with the given hash
+        assert!(blocks.block_number_from_block_id(BlockId::Hash(Felt::from(1).into())).is_none());
+        assert!(
+            blocks
+                .block_number_from_block_id(BlockId::Tag(starknet_rs_core::types::BlockTag::Latest))
+                .is_some()
+        );
+        assert!(
+            blocks
+                .block_number_from_block_id(BlockId::Tag(
+                    starknet_rs_core::types::BlockTag::Pending
+                ))
+                .is_some()
+        );
+        assert!(blocks.block_number_from_block_id(BlockId::Hash(block_hash.into())).is_some());
+    }
+
+    #[test]
+    fn get_blocks_with_filter() {
+        let mut blocks = StarknetBlocks::default();
+
```

Comment:
asserting length is not enough, we would need to assert real data like hashes or something

Reply:
After this comment many changes were pushed. Can you rereview it

Reply:
There is already such test https://github.com/0xSpaceShard/starknet-devnet-rs/blob/6c53ed1703b03efd59eeae5ba41b40c86081b95c/crates/starknet/src/blocks/mod.rs#L234